src/sentry/models/outbox.py

```diff
@@ -524,28 +521,45 @@ def save(self, **kwds: Any) -> None:  # type: ignore[override]
         metrics.incr("outbox.saved", 1, tags=tags)
         super().save(**kwds)
 
+    def lock_id(self, attrs: Iterable[str]) -> int:
+        # 64 bit integer that roughly encodes a unique, serializable lock identifier
+        return mmh3.hash64(".".join(str(getattr(self, attr)) for attr in attrs))[0]
+
     @contextlib.contextmanager
     def process_shard(
         self, latest_shard_row: OutboxBase | None
     ) -> Generator[OutboxBase | None, None, None]:
         flush_all: bool = not bool(latest_shard_row)
-        next_shard_row: OutboxBase | None
         using: str = db.router.db_for_write(type(self))
-        with transaction.atomic(using=using), django_test_transaction_water_mark(using=using):
-            try:
-                next_shard_row = (
-                    self.selected_messages_in_shard(latest_shard_row=latest_shard_row)
-                    .select_for_update(nowait=flush_all)
-                    .first()
-                )
-            except OperationalError as e:
-                if "LockNotAvailable" in str(e):
-                    # If a non task flush process is running already, allow it to proceed without contention.
-                    next_shard_row = None
-                else:
-                    raise e
 
-            yield next_shard_row
+        shard_lock_id = self.lock_id(self.sharding_columns)
+        obtained_lock = True
+
+        try:
+            with connections[using].cursor() as cursor:
+                if flush_all:
```

Comment:
In the flush_all case, we do not want to block -- we try to obtain the shard, but will happily skip this shard if it is already being produced elsewise.